File: packages/host-contract/src/legacy-mapping.ts
/**
 * GLM-CAL-03: pure mapping from the legacy assignment record shape
 * (`TaskRunInfo` in `types.ts`) to the canonical WorkItem status axes defined
 * by the control-plane plan (Sections 5.1/5.2).
 *
 * The mapping is pure: no I/O, no network, no database access, deterministic.
 * Unknown legacy statuses are rejected with a typed error, never silently
 * mapped. Legacy IDs are preserved verbatim in `legacy_compat_id`; no durable
 * `work_item_id` is minted here — that is the CP-20 migration runner's job.
 */

/** Canonical work status values (control-plane plan, Section 5.1). */
export const WORK_STATUSES = [
  "backlog",
  "todo",
  "in_progress",
  "in_review",
  "blocked",
  "done",
  "cancelled",
] as const;
export type WorkStatus = (typeof WORK_STATUSES)[number];

/** Canonical execution phase values (control-plane plan, Section 5.2). */
export const EXECUTION_PHASES = [
  "none",
  "queued",
  "planning",
  "awaiting_plan_approval",
  "running",
  "awaiting_input",
  "awaiting_approval",
  "verifying",
  "reviewing",
  "retrying",
  "paused",
  "failed",
  "needs_attention",
  "terminal",
] as const;
export type ExecutionPhase = (typeof EXECUTION_PHASES)[number];

/** Canonical WorkItem draft produced by the mapping. Not a durable record:
 * `work_item_id` is always `null` because this pure function never invents
 * IDs — the CP-20 migration runner assigns durable IDs. */
export type CanonicalWorkItemDraft = {
  work_item_id: string | null;
  title: string;
  work_status: WorkStatus;
  execution_phase: ExecutionPhase;
  legacy_compat_id: string;
  created_at: string;
};

export type LegacyMappingErrorKind =
  | "missing_required_field"
  | "unknown_legacy_status"
  | "invalid_legacy_id";

/** Typed mapping error. Malformed input is always rejected, never silently mapped. */
export class LegacyMappingError extends Error {
  constructor(
    public readonly kind: LegacyMappingErrorKind,
    public readonly field?: string,
    public readonly value?: unknown,
  ) {
    super(field ? `${kind}: ${field}` : kind);
    this.name = "LegacyMappingError";
  }
}

/**
 * Status mapping (amended 2026-08-03): legacy `failed` maps to work_status
 * `in_progress` + execution_phase `failed` — the work was started but is not
 * complete; attention is derived from the execution phase, not WorkStatus.
 */
const STATUS_MAPPING: Record<string, { work_status: WorkStatus; execution_phase: ExecutionPhase }> = {
  queued: { work_status: "todo", execution_phase: "queued" },
  running: { work_status: "in_progress", execution_phase: "running" },
  succeeded: { work_status: "done", execution_phase: "terminal" },
  failed: { work_status: "in_progress", execution_phase: "failed" },
  cancelled: { work_status: "cancelled", execution_phase: "terminal" },
};

function requiredString(candidate: Record<string, unknown>, key: string, field: string): string {
  const value = candidate[key];
  if (typeof value !== "string") {
    throw new LegacyMappingError("missing_required_field", field, value);
  }
  return value;
}

/** Maps one legacy assignment record to the canonical WorkItem status axes.
 * Pure: same input always produces the same output, with no side effects. */
export function mapLegacyAssignment(input: unknown): CanonicalWorkItemDraft {
  if (typeof input !== "object" || input === null || Array.isArray(input)) {
    throw new LegacyMappingError("missing_required_field", "assignment", input);
  }
  const candidate = input as Record<string, unknown>;

  const idValue = candidate.id;
  if (idValue === null || idValue === undefined) {
    throw new LegacyMappingError("missing_required_field", "id");
  }
  if (typeof idValue !== "string" || idValue.length === 0) {
    throw new LegacyMappingError("invalid_legacy_id", "id", idValue);
  }

  const title = requiredString(candidate, "title", "title");
  const statusValue = candidate.status;
  if (statusValue === null || statusValue === undefined) {
    throw new LegacyMappingError("missing_required_field", "status");
  }
  if (typeof statusValue !== "string") {
    throw new LegacyMappingError("missing_required_field", "status", statusValue);
  }
  const createdAt = requiredString(candidate, "createdAt", "created_at");

  const mapped = STATUS_MAPPING[statusValue];
  if (!mapped) {
    throw new LegacyMappingError("unknown_legacy_status", "status", statusValue);
  }

  return {
    work_item_id: null,
    title,
    work_status: mapped.work_status,
    execution_phase: mapped.execution_phase,
    legacy_compat_id: idValue,
    created_at: createdAt,
  };
}

File: src-tauri/crates/altai-core/src/legacy_mapping.rs
//! GLM-CAL-03: pure mapping from the legacy assignment record shape
//! (`TaskRunInfo` in `packages/host-contract/src/types.ts`) to the canonical
//! WorkItem status axes defined by the control-plane plan (Sections 5.1/5.2).
//!
//! The mapping is pure: no I/O, no network, no database access, deterministic.
//! Unknown legacy statuses are rejected with a typed error, never silently
//! mapped. Legacy IDs are preserved verbatim in `legacy_compat_id`; no durable
//! `work_item_id` is minted here — that is the CP-20 migration runner's job.

use std::fmt;

/// Canonical work status values (control-plane plan, Section 5.1).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum WorkStatus {
    Backlog,
    Todo,
    InProgress,
    InReview,
    Blocked,
    Done,
    Cancelled,
}

impl WorkStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Backlog => "backlog",
            Self::Todo => "todo",
            Self::InProgress => "in_progress",
            Self::InReview => "in_review",
            Self::Blocked => "blocked",
            Self::Done => "done",
            Self::Cancelled => "cancelled",
        }
    }
}

impl fmt::Display for WorkStatus {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// Canonical execution phase values (control-plane plan, Section 5.2).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExecutionPhase {
    None,
    Queued,
    Planning,
    AwaitingPlanApproval,
    Running,
    AwaitingInput,
    AwaitingApproval,
    Verifying,
    Reviewing,
    Retrying,
    Paused,
    Failed,
    NeedsAttention,
    Terminal,
}

impl ExecutionPhase {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::None => "none",
            Self::Queued => "queued",
            Self::Planning => "planning",
            Self::AwaitingPlanApproval => "awaiting_plan_approval",
            Self::Running => "running",
            Self::AwaitingInput => "awaiting_input",
            Self::AwaitingApproval => "awaiting_approval",
            Self::Verifying => "verifying",
            Self::Reviewing => "reviewing",
            Self::Retrying => "retrying",
            Self::Paused => "paused",
            Self::Failed => "failed",
            Self::NeedsAttention => "needs_attention",
            Self::Terminal => "terminal",
        }
    }
}

impl fmt::Display for ExecutionPhase {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// Legacy assignment record as stored in altai-assignments.json (the
/// `TaskRunInfo` concept). `Option` fields model JSON `null` or absent keys;
/// both are rejected as `MissingRequiredField`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LegacyAssignment {
    pub id: Option<String>,
    pub title: Option<String>,
    pub status: Option<String>,
    pub created_at: Option<String>,
}

/// Canonical WorkItem draft produced by the mapping. This is not a durable
/// record: `work_item_id` is always `None` because this pure function never
/// invents IDs — the CP-20 migration runner assigns durable IDs.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CanonicalWorkItemDraft {
    pub work_item_id: Option<String>,
    pub title: String,
    pub work_status: WorkStatus,
    pub execution_phase: ExecutionPhase,
    pub legacy_compat_id: String,
    pub created_at: String,
}

/// Typed mapping errors. Malformed input is always rejected, never silently
/// mapped.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum LegacyMappingError {
    MissingRequiredField(&'static str),
    UnknownLegacyStatus(String),
    InvalidLegacyId,
}

impl fmt::Display for LegacyMappingError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::MissingRequiredField(field) => {
                write!(f, "missing required field {field:?}")
            }
            Self::UnknownLegacyStatus(status) => {
                write!(f, "unknown legacy status {status:?}")
            }
            Self::InvalidLegacyId => write!(f, "legacy id is empty"),
        }
    }
}

impl std::error::Error for LegacyMappingError {}

/// Maps one legacy assignment record to the canonical WorkItem status axes.
///
/// Status mapping (amended 2026-08-03):
/// - `queued`    -> work_status `todo`        + execution_phase `queued`
/// - `running`   -> work_status `in_progress` + execution_phase `running`
/// - `succeeded` -> work_status `done`        + execution_phase `terminal`
/// - `failed`    -> work_status `in_progress` + execution_phase `failed`
/// - `cancelled` -> work_status `cancelled`   + execution_phase `terminal`
///
/// Pure: same input always produces the same output, with no side effects.
pub fn map_legacy_assignment(
    input: &LegacyAssignment,
) -> Result<CanonicalWorkItemDraft, LegacyMappingError> {
    let id = input
        .id
        .as_deref()
        .ok_or(LegacyMappingError::MissingRequiredField("id"))?;
    if id.is_empty() {
        return Err(LegacyMappingError::InvalidLegacyId);
    }
    let title = input
        .title
        .clone()
        .ok_or(LegacyMappingError::MissingRequiredField("title"))?;
    let status = input
        .status
        .as_deref()
        .ok_or(LegacyMappingError::MissingRequiredField("status"))?;
    let created_at = input
        .created_at
        .clone()
        .ok_or(LegacyMappingError::MissingRequiredField("created_at"))?;

    let (work_status, execution_phase) = match status {
        "queued" => (WorkStatus::Todo, ExecutionPhase::Queued),
        "running" => (WorkStatus::InProgress, ExecutionPhase::Running),
        "succeeded" => (WorkStatus::Done, ExecutionPhase::Terminal),
        "failed" => (WorkStatus::InProgress, ExecutionPhase::Failed),
        "cancelled" => (WorkStatus::Cancelled, ExecutionPhase::Terminal),
        other => return Err(LegacyMappingError::UnknownLegacyStatus(other.to_string())),
    };

    Ok(CanonicalWorkItemDraft {
        work_item_id: None,
        title,
        work_status,
        execution_phase,
        legacy_compat_id: id.to_string(),
        created_at,
    })
}
